src/Cortside.RestApiClient/RestApiClient.cs

```diff
@@ -0,0 +1,208 @@
+            if (options.XmlSerializer) {
+                client.UseXml();
+            }
+        }
+
+        public IDistributedCache Cache {
+            get { return options.Cache; }
+        }
+
+        public Uri BuildUri(IRestApiRequest request) => client.BuildUri(request.RestRequest);
+
+        private void TimeoutCheck(IRestApiRequest request, RestResponse response) {
+            if (response.ResponseStatus == ResponseStatus.TimedOut) {
+                response.ErrorMessage = null;
+                response.ErrorException = null;
+                LogError(request, response);
+
+                if (options.ThrowOnAnyError) {
+                    var timeouts = new int[] { request.Timeout, client.Options.MaxTimeout };
+                    var timeout = timeouts.Where(x => x > 0).Min();
+                    throw new TimeoutException($"Timeout of {timeout} exceeded");
+                }
+            }
+        }
+
+        protected async Task<RestResponse> InnerExecuteAsync(IRestApiRequest request) {
+            var policy = request.Policy ?? options.Policy;
+
+            var correlationId = CorrelationContext.GetCorrelationId();
+            request.AddHeader("Request-Id", correlationId);
+            request.AddHeader("X-Correlation-Id", correlationId);
+
+            int retry = 0;
+            var response = await policy.ExecuteAsync(async () => {
+                // TODO: log the attempt # as property
+                logger.LogInformation($"Attempt {retry++}");
+
+                var response = await client.ExecuteAsync(request.RestRequest).ConfigureAwait(false);
+                logger.LogInformation("Response {retry}: Status Code = {StatusCode} Data = {Content}", retry, response.StatusCode, response.Content);
+                TimeoutCheck(request, response);
+                if (options.ThrowOnAnyError && response.ErrorException != null) {
+                    throw response.ErrorException;
+                }
+                return response;
+            }).ConfigureAwait(false);
+
+            logger.LogDebug("Response to {url} returned with status code {StatusCode} and content: {Content}", client.BuildUri(request.RestRequest), response.StatusCode, response.Content);
+            TimeoutCheck(request, response);
+            var exception = response.ErrorException;
+            if (options.ThrowOnAnyError && exception != null) {
+                throw exception;
+            }
+
+            // TODO: should this check status code? anything else?
+            if ((request.FollowRedirects ?? options.FollowRedirects) && response.Headers?.Any(h => h.Name.Equals("location", StringComparison.InvariantCultureIgnoreCase)) == true) {
+                var url = response.Headers.FirstOrDefault(h => h.Name.Equals("location", StringComparison.InvariantCultureIgnoreCase)).Value.ToString();
+                logger.LogInformation($"Following redirect to {url}");
+                var redirectRequest = new RestApiRequest(url, Method.Get);
```

Comment:
looks right - should have the authenticator (if any) applied on the new request

Reply:
phew!  it _seemed_ right and seemed fairly easy to handle given the way things work now.